src/java.base/share/classes/java/security/Provider.java

```diff
@@ -684,30 +683,737 @@ private void checkInitialized() {
+                    // have a corresponding Properties map entry.
+                    String oldPropKey = serviceProps.remove(aliasKey);
+                    assert oldPropKey != null : "Invalid alias property.";
+                    Provider.super.remove(oldPropKey);
+
+                    return true;
+                }
+                assert false : "Should not reach.";
+                return false;
+            });
+        }
+
+        boolean removeAttributeLegacy(ServiceKey key, String attrName,
+                String attrValue) {
+            assert key != null && attrName != null && attrValue != null :
+                    "Attribute information missing.";
+            return updateSvc(key, (MappingInfo oldMi, Service newSvc) -> {
+                Map<UString, String> oldAttrProps =
+                        serviceAttrProps.get(oldMi.algKey);
+                if (oldAttrProps != null) {
+                    // The service was found and has attributes.
+                    assert oldMi.svc != null :
+                            "Inconsistent service attributes data.";
+
+                    newSvc.removeAttribute(attrName, attrValue);
+                    assert newSvc.getAttribute(attrName) == null :
+                            "Attribute was not removed from the service.";
+
+                    // Update the Properties map to reflect the attribute
+                    // removal. Note: oldPropKey cannot be null because
+                    // attributes always have a corresponding Properties map
+                    // entry.
+                    String oldPropKey = oldAttrProps.remove(
+                            new UString(attrName));
+                    assert oldPropKey != null : "Invalid attribute property.";
+                    Provider.super.remove(oldPropKey);
+
+                    if (oldAttrProps.isEmpty()) {
+                        // If the removed attribute was the last one, remove the
+                        // map.
+                        serviceAttrProps.remove(oldMi.algKey);
+                    }
+
+                    return true;
+                }
+                assert false : "Should not reach.";
+                return false;
+            });
+        }
+
+        @FunctionalInterface
+        private interface ServiceUpdateCallback {
+            boolean apply(MappingInfo oldMi, Service newSvc);
+        }
+
+        /*
+         * This method tries to find a service on the map (based on an
+         * algorithm or alias) and pass a copy of it to an update callback
+         * (copy-on-write). If the service found was added with the Current API,
+         * no update should be done. If a service was not found, a new instance
+         * may be created.
+         *
+         * The updated version of the service is put on the services map.
+         * Algorithm and alias based entries pointing to the old version of the
+         * service are overwritten.
+         */
```

Comment:
Comment for this method should note that this method is only for the Legacy registration...

Reply:
Sounds good. I added a note.